src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#include "precompiled.hpp"
+#include "jvm.h"
+#include "logging/logAsyncFlusher.hpp"
+#include "logging/logConfiguration.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logHandle.hpp"
+#include "runtime/atomic.hpp"
+
+void AsyncLogMessage::writeback() {
+  if (_message != NULL) {
+    _output.write_blocking(_decorations, _message);
+    os::free(_message);
+  }
+}
+
+void LogAsyncFlusher::enqueue_impl(const AsyncLogMessage& msg) {
+  assert_lock_strong(&_lock);
+
+  if (_buffer.size() >= _buffer_max_size)  {
+    const AsyncLogMessage* h = _buffer.front();
+    assert(h != NULL, "sanity check");
+
+    if (h->message() != nullptr) {
+      bool p_created;
+      uint32_t* counter = _stats.add_if_absent(h->output(), 0, &p_created);
+      *counter = *counter + 1;
+
+      if (Verbose) {
+        // Temporarily turn off SerializeVMOutput so defaultStream will not
+        // invoke set_owner(self) for tty_lock.
+        FlagSetting t(SerializeVMOutput, false);
+        // The writing below can not guarantee non-blocking because tty may be piped by the filesystems
+        // or throttled by XOFF, so only dump the dropping message in Verbose mode.
+        tty->print_cr("asynclog dropping message: %s", h->message());
+      }
```

Comment:
I'm not sure this is needed or sensible. If we are flooded, we would then proceed to print every dropped message to tty? So now we flood tty? 

I'd hope exhausting the output buffer is rare, and easily remedied by telling the user to increase the buffer or stop using async mode.